client/app/house/hooks/useCreateUserHouseOnMount.ts

```diff
@@ -0,0 +1,63 @@
+import { useState, useRef, useEffect } from 'react'
+import { useRouter, useSearchParams } from 'next/navigation'
+import { useQueryClient } from '@tanstack/react-query'
+import { INVITED_HOUSE_ID } from '@/_consts'
+import { useGetHouseInfo } from '@/_hooks'
+import { useCreateUserHouseMutation } from '@/_hooks/api'
+import { toast } from '@/_hooks/useToast'
+import { isErrorWithMessage } from '@/_utils'
+
+export const useCreateUserHouseOnMount = () => {
+  const router = useRouter()
+  const searchParams = useSearchParams()
+  const invitedHouseId = searchParams.get(INVITED_HOUSE_ID)
+
+  const queryClient = useQueryClient()
+
+  const [isLoading, setIsLoading] = useState(true)
+  const hasMutated = useRef(false)
+
+  const { houseId } = useGetHouseInfo()
+  const { mutateAsync } = useCreateUserHouseMutation()
+
+  // NOTE mutate function does not work inside of useEffect for onMount because of strict mode (it runs twice)
+  const mutateUserHouse = async (invitedHouseId: string) => {
+    try {
+      await mutateAsync(invitedHouseId)
+      queryClient.invalidateQueries({ queryKey: ['me'] })
+      router.push('/')
+    } catch (error) {
+      let message = 'Please try again later.'
+
+      if (isErrorWithMessage(error) && error.response) {
+        message = error.response.data.message
+      }
+
+      toast({
+        variant: 'destructive',
+        title: 'Failed to create house..',
+        description: message,
+      })
+
+      router.push('/house/create')
+    } finally {
+      setIsLoading(false)
+    }
+  }
+
+  useEffect(() => {
+    if (houseId) {
+      router.push('/')
+    }
+
+    // If user had invitedHouse Id then create user house before showing house setting page
+    if (invitedHouseId && !houseId && !hasMutated.current) {
+      mutateUserHouse(invitedHouseId)
+      hasMutated.current = true
```

Comment:
I assume that you use useRef to prevent running `mutateUserHouse(invitedHouseId)` twice in strict mode, but when you call `mutateUserHouse(invitedHouseId)` ,
```
await mutateAsync(invitedHouseId)
queryClient.invalidateQueries({ queryKey: ['me'] })
router.push('/')
```
 might run before `hasMutated.current = true` and `hasMutated.current = true` might end up not being executed since router.push causes this component to be unmounted first.

So do you think it's better to call `hasMutated.current = true` right after `await mutateAsync(invitedHouseId)` so that it's guratneed that `mutateUserHouse(invitedHouseId)` is only executed once?

Reply:
Thanks for your insight! But in this case, I don't want to put after `await mutateAsync(invitedHouseId)` because I don't want to await. Mayebe it made you confused so I changed the place before `mutateUserHouse(invitedHouseId)`.https://github.com/HAK2024/fairy-share/pull/89/commits/fea436a84a5d956425088e464342a9b201e097ee

 If you wait until `mutateAsync` is finished, the function of `mutateUserHouse` runs twice because of the strict mode. I can even add clean up function too because it runs twice...

Reply:
Oh I see! Are you planning to add clean up fucntion then?

Btw, how `mutateAsync` and `mutate` changes the result in this code? Because I thougt mutate is also async function

Reply:
> Are you planning to add clean up fucntion then?

No, sorry for the confusion but I don't need to add clean up function in this case because of await router.push. 

>how mutateAsync and mutate changes the result in this code? 
Because I thougt mutate is also async function

Becasue I wanted to use try and catch inside of useEffect. The call back of `onError` or any status provided by react query does not work inside of useEffect onMount because I'm controlling with `hasMutated.current`. 

https://tanstack.com/query/latest/docs/framework/react/guides/mutations#promises

Reply:
Thanks for the explnation! That makes sense!